### ISSUE TITLE
References to Multiple Private QT Headers

### ISSUE THREAD

**Issue by DavidMerzJr** [swri-robotics/ros_scxml#23]
This library makes direct use of private QT headers.  These headers are not meant to be used outside of the QT libraries and are not exported/included in standard APT packages.  This makes it difficult to properly build this package, even when the required QT version already exists on the machine.

https://github.com/swri-robotics/ros_scxml/blob/master/scxml_core/include/scxml_core/state_machine.h#L45

For example, Ubuntu 20.04 has libqt5core5a available at QT version 5.12.8.  This should be more than enough to build this library - except that this library uses private headers are not included.  As a result, to use this package, we have to do a custom install (of a QT version that is already installed) and then change environment variables.  Changing those environment variables causes conflicts with other projects unless every possible QT module is also built from source.

Refactoring this library to avoid use of private headers would vastly improve its ease of installation and use.

**Comment by jrgnicho** [comment]
I'm on board with removing the dependency on those private headers, the challenge is that a lot of the functionality that this library uses isn't publicly exposed.  One option would be to copy those private headers into this repo although this has the downside that improvements made to the code in these headers won't be used by ros_scxml anymore

**Comment by Levi-Armstrong** [comment]
Going to try and leverage CMake External project functionality to pull the source for building.

**Comment by marip8** [comment]
From what I can tell, these are the differences between what this library provides through the use of the private headers and what comes standard with the `QtScxmlStateMachine` class:

## Get all available transitions for a given state
This library allows you to query what transitions are available at the current state. The only use-case that I have come across for this is enabling/disabling GUI buttons that are used to issue state machine transition commands. I'm not sure what programmatic use this capability really serves because the state machine code is written with the knowledge of the state machine itself (i.e. every state should already know what transitions it has available without having to ask).

There might also be a way to glean this information by parsing the .scxml file directly or leveraging some of the Qt objects that parse that file into a `QScxmlStateMachine` object

## Pre-condition callbacks
Currently "pre-condition" callbacks are used to determine whether it is valid to transition from one state to another state. If the callback fails, the transition does not occur. For the most part I think you can design your state machine or its state entry callbacks such that pre-condition callbacks are not necessary

## Restriction of transitions until asynchronous callback function has finished
I believe the current implementation keeps track of the `future` objects produced by asynchronous on-entry callbacks. It may or may not allow the state machine to issue a transition command until that callback has finished (or queues up events to execute once it is finished). I think this functionality is independent of the code in the private Qt headers and could be replicated in a state machine class based only off of the public API

## TLDR
- For relatively simple state machines that don't require asynchronous callbacks, it's probably easier to use the `QScxmlStateMachine` class directly (i.e. public Qt API). Coincidentally, I made the switch on a project today and it was actually very easy because I wasn't even using any of the special features from the private Qt API
- I think we could eliminate the concept of pre-condition callbacks, remove the ability to get all transitions for a particular state, and transition the asynchronous callback capability into a new state machine class that doesn't depend on private headers

**Comment by DavidMerzJr** [comment]
@marip8 Thanks for taking the time to dig through this code more deeply and find the actual places where the private headers are needed.  I agree that the additional functionality from those headers does not seem to be very great or critical.  I would be happy to support an effort to implement those changes, as the custom QT version/installation is currently a strong disincentive (unincentive? uncentive? discentive?) to using this library.

**Comment by Levi-Armstrong** [comment]
I posted an issue on Qt and they recommended taking a look at the library below. It provides the same functionality and they have several other tool which look nice for the QtScxml state machine. This can also be install using apt on ubuntu. What do you all think about using this?

https://github.com/KDAB/GammaRay/blob/5.15/plugins/statemachineviewer/statemachinedebuginterface.h

**Comment by Levi-Armstrong** [comment]
Well it looks like the code is GPL.  :(

**Comment by marip8** [comment]
> I would be happy to support an effort to implement those changes, as the custom QT version/installation is currently a strong disincentive (unincentive? uncentive? discentive?) to using this library.

You might consider using the `QScxmlStateMachine` class directly. I would only consider writing a revised wrapper for this class if we really needed to do something special with long-running asynchronous tasks in entry callbacks. For the most part, I think you could write the asynchronous function in such a way that you could spin it off and not necessarily need to track when it finishes

> What do you all think about using this?
>
> https://github.com/KDAB/GammaRay/blob/5.15/plugins/statemachineviewer/statemachinedebuginterface.h

I've used GammaRay before as a debugging tool and it's really helpful. I've never thought about integrating that GUI into an application though.

**Comment by marip8** [comment]
I did a little more digging, and it appears that [Qt SCXML implements the SCXML specification](https://doc.qt.io/qt-5/qtscxml-overview.html) which is defined [here](https://www.w3.org/TR/scxml/). According to the specification, the XML file must have [state tags](https://www.w3.org/TR/scxml/#state) which include an `id` attribute and [`transition` child element](https://www.w3.org/TR/scxml/#transition). The `transition` element has an `event` attribute (i.e. the name of the transition) and a `target` attribute (i.e. the name of the state to which the transition will move the state machine when invoked).

We should be able to write a very simple XML parser that can construct a map of known states and valid transitions from the SCXML file directly. The `QStateMachine` class doesn't give very good feedback if you attempt to connect states that don't exist or with transitions that don't exist/aren't valid, so having a map like this would be good for validating operations like the addition of state entry callbacks and the invocation of transitions

**Comment by jrgnicho** [comment]
The scxml format allows defining conditional expressions that are evaluated prior to executing a transition.  The  only way to check if the current value of these conditional expressions is through the private headers so having just a simple transition table map won't allow us to check for that.

**Comment by jrgnicho** [comment]
I think that leveraging the CMake external project feature as @Levi-Armstrong  would be the best option although I'm not sure whether or not that is doable

**Comment by DavidMerzJr** [comment]
> The scxml format allows defining conditional expressions that are evaluated prior to executing a transition. The only way to check if the current value of these conditional expressions is through the private headers so having just a simple transition table map won't allow us to check for that.

Couldn't a user just put the conditional expression in their calling C++ code before attempting to execute a transition?

Edit:  But also, why would the calling code need to be able to check the value of those expressions?  Isn't making the calling code somewhat state-machine-blind sort of the point of the whole library in the first place?

**Comment by Levi-Armstrong** [comment]
If @marip8 method works then I believe it would be a more desirable because it is using public api with will not change within a major release of Qt. The other approach will be fragile. We may have time to test this out on a project.

**Comment by marip8** [comment]
> The scxml format allows defining conditional expressions that are evaluated prior to executing a transition

I did notice this during testing with the PackML state machine. I think you could argue that the transition is still defined for that particular state, but the execution of that transition may not be valid at run time. From that perspective, creating a map of states and transitions would still be a valid approach, and you would still have the tools to query what state the state machine is in and what transitions are nominally available.

We might also be able to find the class or functions that Qt is using to compile those conditional expressions. If those exist in the public API, we could include them as part of our state machine wrapper

**Comment by marip8** [comment]
Closing; addressed by #30
